jetty-core/jetty-http2/jetty-http2-client-transport/src/main/java/org/eclipse/jetty/http2/client/transport/internal/HttpReceiverOverHTTP2.java

```diff
@@ -80,10 +81,21 @@ public Content.Chunk read(boolean fillInterestIfNeeded)
             return null;
         }
         DataFrame frame = data.frame();
-        boolean last = frame.remaining() == 0 && frame.isEndStream();
-        if (last)
-            responseSuccess(getHttpExchange(), null);
-        return Content.Chunk.from(frame.getData(), last, data);
+        ByteBuffer buffer = frame.getData();
+        if (!buffer.hasRemaining())
+        {
+            data.release();
+            if (frame.isEndStream())
+            {
+                responseSuccess(getHttpExchange(), null);
+                return Content.Chunk.EOF;
+            }
+            else
+            {
+                return Content.Chunk.EMPTY;
+            }
+        }
+        return Content.Chunk.from(buffer, false, data);
```

Comment:
Doesn't `from` now do all this logic about choosing to return `EMPTY` or `EOF` if possible?  It was much simpler code to just call `from` and let it work out the details, rather than make the caller do this all the time.

Reply:
That can be done with the 3 variants of `from` that take a `ByteBuffer`-controlling object as a parameter (when `from` wants to replace the empty byte buffer with EOF/EMPTY, it must release the byte buffer) but not with the 4th one.

I've changed the replacement code scattered everywhere so that this is now done by the 3 `from` variants, I made the 4th one throw an exception and changed the calls to the 4th one to one of the 3 non-throwing one where appropriate.

I find this much cleaner, so I'm in favor of this change.